proposals/0003-health-checks.md

```diff
@@ -0,0 +1,103 @@
+# Service Healthchecks
+
+* Proposal: [MP-0003](0003-health-checks.md)
+* Authors: [Clement Escoffier](https://github.com/cescoffier), [Heiko Rupp](https://github.com/pilhuhn), [Heiko Braun] (https://github.com/heiko-braun)
+
+* Status: **Awaiting review**
+
+*During the review process, add the following fields as needed:*
+
+* Decision Notes: [Discussion thread topic covering the  Rationale](https://groups.google.com/forum/#!forum/microprofile), [Discussion thread topic with additional Commentary](https://groups.google.com/forum/#!forum/microprofile)
+
+## Introduction
+
+This document describes a protocol (wireformat, semantics and possible forms of interactions) between system components that need to determine the “liveliness” of computing nodes in a bigger architecture.
+
+
+Mailinglist thread: [Discussion thread topic for that proposal](https://groups.google.com/forum/#!topic/microprofile/jIZAKiu76ys)
+
+## Motivation
+
+Health checks are used to probe the state of a computing node from another other machine (i.e. kubernetes service controller) with the primary target being cloud infrastructure environments where automated processes maintain the state of computing nodes. In this scenario health checks are used to determine if a computing node needs to be discarded (terminated, shutdown) and eventually replaced by another (healthy) instance.
+
+It’s not intended (although could be used) as a monitoring solution for human operators.
+
+## Proposed solution
+
+The proposed solution breaks down into two parts:
+
+- A health checks protocol and wireformat description
+- A Java API to implement health check procedures
+
+## Detailed design
+
+### Protocol
+
+For a detailed description of the protocol see the companion document that the describes [the protocol and wireformat](0003-spec.md)
+
+### Java API
+
+The main API to describe health check results (responses) is the `HealthStatus` interface:
+
+```
+File path = new File(System.getProperty("user.home"));
+long freeBytes = path.getFreeSpace();
+long threshold = 1024 * 1024 * 100; // 100mb
+return freeBytes>threshold ?
+        HealthStatus.
+                named("diskspace")
+                .up()
+                .withAttribute("freebytes", freeBytes) :
+        HealthStatus.
+                named("diskspace")
+                .down()
```

Comment:
I think we could have another state called .warning("message"). We could have a resource which is not DOWN yet but can be soon if not fixed. One example can be the disk health check, if disk is over 90% of usage it can be a problem soon. So, better to have a warning before fail than having a surprise when is down.